File: Golang/10-Hashing/hashing.go
// 10 - Hashing: a hash map built from scratch (separate chaining + rehashing),
// plus the patterns hashing exists to serve, and an LRU cache.
//
// Run:  go run hashing.go
package main

import (
	"container/list"
	"fmt"
	"sort"
)

// ============================================================================
// 1. A hash map from scratch (separate chaining)
// ============================================================================

type entry struct {
	key   string
	value int
}

// HashMap is a string -> int dictionary over an array of buckets, each holding
// a chain. Go's real map uses 8-slot buckets with overflow chaining, but the
// moving parts are the same: hash, index, collide, load factor, resize.
type HashMap struct {
	buckets [][]entry
	size    int
}

const (
	initialCapacity = 8
	maxLoadFactor   = 0.75
)

func NewHashMap() *HashMap {
	return &HashMap{buckets: make([][]entry, initialCapacity)}
}

// fnv1a is a small, well-distributed string hash. Go's runtime uses a
// randomly seeded hash instead, which is what defends against hash flooding.
func fnv1a(s string) uint64 {
	const (
		offset = 14695981039346656037
		prime  = 1099511628211
	)
	h := uint64(offset)
	for i := 0; i < len(s); i++ {
		h ^= uint64(s[i])
		h *= prime
	}
	return h
}

func (m *HashMap) indexFor(key string) int {
	return int(fnv1a(key) % uint64(len(m.buckets)))
}

func (m *HashMap) Len() int            { return m.size }
func (m *HashMap) Capacity() int       { return len(m.buckets) }
func (m *HashMap) LoadFactor() float64 { return float64(m.size) / float64(len(m.buckets)) }

// Put is O(1) average, amortised across the occasional resize.
func (m *HashMap) Put(key string, value int) {
	i := m.indexFor(key)
	for j := range m.buckets[i] {
		if m.buckets[i][j].key == key { // equal keys overwrite
			m.buckets[i][j].value = value
			return
		}
	}
	m.buckets[i] = append(m.buckets[i], entry{key: key, value: value})
	m.size++
	if m.LoadFactor() > maxLoadFactor {
		m.resize()
	}
}

// Get returns (value, found). O(1) average, O(chain) worst case.
func (m *HashMap) Get(key string) (int, bool) {
	for _, e := range m.buckets[m.indexFor(key)] {
		if e.key == key {
			return e.value, true
		}
	}
	return 0, false
}

func (m *HashMap) Delete(key string) bool {
	i := m.indexFor(key)
	for j, e := range m.buckets[i] {
		if e.key == key {
			// Remove index j while preserving the rest of the chain.
			m.buckets[i] = append(m.buckets[i][:j], m.buckets[i][j+1:]...)
			m.size--
			return true
		}
	}
	return false
}

// resize doubles the capacity and REHASHES everything: the bucket index
// depends on the capacity, so every key belongs somewhere new. O(n).
func (m *HashMap) resize() {
	old := m.buckets
	m.buckets = make([][]entry, len(old)*2)
	m.size = 0
	for _, chain := range old {
		for _, e := range chain {
			m.Put(e.key, e.value)
		}
	}
}

func (m *HashMap) Keys() []string {
	keys := make([]string, 0, m.size)
	for _, chain := range m.buckets {
		for _, e := range chain {
			keys = append(keys, e.key)
		}
	}
	return keys
}

// LongestChain is a diagnostic: how badly is this table colliding?
func (m *HashMap) LongestChain() int {
	worst := 0
	for _, chain := range m.buckets {
		worst = max(worst, len(chain))
	}
	return worst
}

// ============================================================================
// 2. Frequency map
// ============================================================================

// CharFrequency counts bytes. m[c]++ works on a missing key because the read
// returns the zero value - no initialisation check needed.
func CharFrequency(s string) map[byte]int {
	freq := make(map[byte]int, len(s))
	for i := 0; i < len(s); i++ {
		freq[s[i]]++
	}
	return freq
}

func FirstUniqueChar(s string) int {
	freq := CharFrequency(s)
	for i := 0; i < len(s); i++ {
		if freq[s[i]] == 1 {
			return i
		}
	}
	return -1
}

// SortedKeys exists because Go randomises map iteration order on purpose.
// Anything deterministic must sort first.
func SortedKeys(m map[string]int) []string {
	keys := make([]string, 0, len(m))
	for k := range m {
		keys = append(keys, k)
	}
	sort.Strings(keys)
	return keys
}

// ============================================================================
// 3. Complement lookup
// ============================================================================

// TwoSum returns the indices of the pair summing to target, or (-1,-1).
// Brute force asks "does x pair with a later element?" - O(n^2). Hashing
// flips it to "have I already seen the complement?" - O(1) per check.
func TwoSum(nums []int, target int) (int, int) {
	seen := make(map[int]int, len(nums)) // value -> index
	for i, x := range nums {
		if j, ok := seen[target-x]; ok {
			return j, i
		}
		seen[x] = i // store AFTER checking, so nothing pairs with itself
	}
	return -1, -1
}

// ContainsDuplicate uses map[T]struct{} - the idiomatic Go set, 0 bytes per value.
func ContainsDuplicate(nums []int) bool {
	seen := make(map[int]struct{}, len(nums))
	for _, x := range nums {
		if _, dup := seen[x]; dup {
			return true
		}
		seen[x] = struct{}{}
	}
	return false
}

// ============================================================================
// 4. Grouping by a computed key
// ============================================================================

// GroupAnagrams keys on a [26]int count array. Arrays ARE comparable in Go,
// so they can be map keys directly - no string encoding needed. Slices cannot.
func GroupAnagrams(words []string) [][]string {
	groups := make(map[[26]int][]string)
	for _, w := range words {
		var key [26]int
		for i := 0; i < len(w); i++ {
			key[w[i]-'a']++
		}
		groups[key] = append(groups[key], w)
	}
	out := make([][]string, 0, len(groups))
	for _, g := range groups {
		out = append(out, g)
	}
	return out
}

// ============================================================================
// 5. Prefix sum + map
// ============================================================================

// SubarraySumEqualsK counts contiguous subarrays summing to k. O(n).
//
// prefix[j] - prefix[i] == k means the subarray (i, j] sums to k, so at each j
// we count earlier prefixes equal to prefix[j] - k. counts[0] = 1 seeds the
// empty prefix, which is what lets subarrays starting at index 0 count.
func SubarraySumEqualsK(nums []int, k int) int {
	counts := map[int]int{0: 1}
	prefix, total := 0, 0
	for _, x := range nums {
		prefix += x
		total += counts[prefix-k] // missing key reads as 0
		counts[prefix]++
	}
	return total
}

// ============================================================================
// 6. Seen set
// ============================================================================

// LongestConsecutive is O(n), not O(n log n): a run is only counted from its
// smallest member, so each run is walked exactly once.
func LongestConsecutive(nums []int) int {
	unique := make(map[int]struct{}, len(nums))
	for _, x := range nums {
		unique[x] = struct{}{}
	}
	best := 0
	for x := range unique {
		if _, hasPrev := unique[x-1]; hasPrev {
			continue // not the start of a run
		}
		length := 1
		for {
			if _, ok := unique[x+length]; !ok {
				break
			}
			length++
		}
		best = max(best, length)
	}
	return best
}

func Intersection(a, b []int) []int {
	lookup := make(map[int]struct{}, len(b))
	for _, x := range b {
		lookup[x] = struct{}{}
	}
	found := make(map[int]struct{})
	for _, x := range a {
		if _, ok := lookup[x]; ok {
			found[x] = struct{}{}
		}
	}
	out := make([]int, 0, len(found))
	for x := range found {
		out = append(out, x)
	}
	sort.Ints(out) // map order is random: sort for a deterministic result
	return out
}

// ============================================================================
// 7. Hashing + ordering: an LRU cache
// ============================================================================

// LRUCache needs BOTH a hash map (O(1) lookup) and a doubly linked list
// (recency order). container/list provides O(1) MoveToFront, and the map
// stores *list.Element pointers into it.
type LRUCache struct {
	capacity int
	order    *list.List            // front = most recently used
	index    map[int]*list.Element // key -> node holding a pair
}

type lruPair struct{ key, value int }

func NewLRUCache(capacity int) (*LRUCache, error) {
	if capacity <= 0 {
		return nil, fmt.Errorf("capacity must be positive, got %d", capacity)
	}
	return &LRUCache{
		capacity: capacity,
		order:    list.New(),
		index:    make(map[int]*list.Element, capacity),
	}, nil
}

func (c *LRUCache) Get(key int) int {
	el, ok := c.index[key]
	if !ok {
		return -1
	}
	c.order.MoveToFront(el) // mark as most recently used
	return el.Value.(*lruPair).value
}

func (c *LRUCache) Put(key, value int) {
	if el, ok := c.index[key]; ok {
		el.Value.(*lruPair).value = value
		c.order.MoveToFront(el)
		return
	}
	if len(c.index) == c.capacity {
		oldest := c.order.Back() // least recently used
		c.order.Remove(oldest)
		delete(c.index, oldest.Value.(*lruPair).key)
	}
	c.index[key] = c.order.PushFront(&lruPair{key: key, value: value})
}

// KeysMostRecentFirst is deterministic because it walks the list, not the map.
func (c *LRUCache) KeysMostRecentFirst() []int {
	out := make([]int, 0, len(c.index))
	for el := c.order.Front(); el != nil; el = el.Next() {
		out = append(out, el.Value.(*lruPair).key)
	}
	return out
}

// ============================================================================
// demo
// ============================================================================

func assert(cond bool, msg string) {
	if !cond {
		panic("assertion failed: " + msg)
	}
}

func equal(a, b []int) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

func main() {
	m := NewHashMap()
	m.Put("cat", 1)
	m.Put("dog", 2)
	m.Put("cat", 9) // overwrite, not duplicate
	assert(m.Len() == 2, "size after overwrite")
	v, ok := m.Get("cat")
	assert(v == 9 && ok, "get returns the newest value")
	_, ok = m.Get("missing")
	assert(!ok, "missing key reports false")
	assert(m.Delete("dog") && !m.Delete("dog"), "delete is idempotent-reporting")
	assert(m.Len() == 1, "size after delete")

	// Resizing: force at least one growth and confirm nothing was lost.
	big := NewHashMap()
	for i := 0; i < 100; i++ {
		big.Put(fmt.Sprintf("key%d", i), i)
	}
	assert(big.Len() == 100, "all keys stored")
	assert(big.Capacity() > initialCapacity, "table grew")
	assert(big.LoadFactor() <= maxLoadFactor, "load factor stayed healthy")
	for i := 0; i < 100; i++ {
		v, ok := big.Get(fmt.Sprintf("key%d", i))
		assert(ok && v == i, "every key survived the rehash")
	}
	assert(big.LongestChain() <= 5, "collisions stay tame")
	assert(len(big.Keys()) == 100, "keys listing")

	assert(CharFrequency("aab")['a'] == 2, "frequency")
	assert(FirstUniqueChar("leetcode") == 0, "first unique")
	assert(FirstUniqueChar("aabb") == -1, "no unique char")
	assert(equalStrings(SortedKeys(map[string]int{"b": 1, "a": 2}), []string{"a", "b"}),
		"sorted keys are deterministic")

	i, j := TwoSum([]int{2, 7, 11, 15}, 9)
	assert(i == 0 && j == 1, "two sum")
	i, j = TwoSum([]int{3, 3}, 6)
	assert(i == 0 && j == 1, "two sum with duplicate values")
	i, _ = TwoSum([]int{1, 2}, 99)
	assert(i == -1, "two sum miss")
	assert(ContainsDuplicate([]int{1, 2, 3, 1}), "duplicate found")
	assert(!ContainsDuplicate([]int{1, 2, 3}), "no duplicate")

	groups := GroupAnagrams([]string{"eat", "tea", "tan", "ate", "nat", "bat"})
	sizes := make([]int, 0, len(groups))
	for _, g := range groups {
		sizes = append(sizes, len(g))
	}
	sort.Ints(sizes)
	assert(equal(sizes, []int{1, 2, 3}), "anagram group sizes")

	assert(SubarraySumEqualsK([]int{1, 1, 1}, 2) == 2, "subarray sum")
	assert(SubarraySumEqualsK([]int{1, 2, 3}, 3) == 2, "subarray sum 2")
	assert(SubarraySumEqualsK([]int{1, -1, 0}, 0) == 3, "negatives work")

	assert(LongestConsecutive([]int{100, 4, 200, 1, 3, 2}) == 4, "consecutive run")
	assert(LongestConsecutive(nil) == 0, "empty input")
	assert(equal(Intersection([]int{1, 2, 2, 1}, []int{2, 2}), []int{2}), "intersection")

	lru, err := NewLRUCache(2)
	assert(err == nil, "cache construction")
	lru.Put(1, 100)
	lru.Put(2, 200)
	assert(lru.Get(1) == 100, "hit")
	lru.Put(3, 300) // evicts key 2, the least recently used
	assert(lru.Get(2) == -1, "evicted")
	assert(equal(lru.KeysMostRecentFirst(), []int{3, 1}), "recency order")
	_, err = NewLRUCache(0)
	assert(err != nil, "zero capacity is an error")

	fmt.Println("10-Hashing (Go): all checks passed")
}

func equalStrings(a, b []string) bool {
	if len(a) != len(b) {
		return false
	}
	for i := range a {
		if a[i] != b[i] {
			return false
		}
	}
	return true
}

File: Python/10-Hashing/hashing.py
"""
10 - Hashing: a hash map built from scratch (separate chaining + resizing),
plus the six patterns that hashing exists to serve.

Run:  python hashing.py
"""

from __future__ import annotations

from collections import Counter, OrderedDict, defaultdict
from typing import Any, Iterator


# ============================================================================
# 1. A hash map from scratch (separate chaining)
# ============================================================================
class HashMap:
    """Dictionary built on an array of buckets, each holding a chain.

    Real CPython dicts use open addressing, but chaining is easier to reason
    about and shows every moving part: hashing, bucket indexing, collision
    handling, load factor and resizing.
    """

    _INITIAL_CAPACITY = 8
    _MAX_LOAD_FACTOR = 0.75      # resize past this, before chains get long

    def __init__(self) -> None:
        self._buckets: list[list[tuple[Any, Any]]] = [
            [] for _ in range(self._INITIAL_CAPACITY)
        ]
        self._size = 0

    # ------------------------------------------------------------- internals
    def _index(self, key: Any) -> int:
        """Map a key to a bucket. hash() can be negative, so mask it first."""
        return (hash(key) & 0x7FFFFFFF) % len(self._buckets)

    def _resize(self) -> None:
        """Double the capacity and REHASH everything. O(n).

        Rehashing is mandatory: the bucket index depends on the capacity, so
        every key belongs somewhere new after the table grows.
        """
        old_buckets = self._buckets
        self._buckets = [[] for _ in range(len(old_buckets) * 2)]
        self._size = 0
        for chain in old_buckets:
            for key, value in chain:
                self.put(key, value)

    @property
    def load_factor(self) -> float:
        return self._size / len(self._buckets)

    @property
    def capacity(self) -> int:
        return len(self._buckets)

    # ---------------------------------------------------------------- public
    def put(self, key: Any, value: Any) -> None:
        """O(1) average, amortised across the occasional resize."""
        chain = self._buckets[self._index(key)]
        for i, (existing_key, _) in enumerate(chain):
            if existing_key == key:            # equal keys must overwrite
                chain[i] = (key, value)
                return
        chain.append((key, value))
        self._size += 1
        if self.load_factor > self._MAX_LOAD_FACTOR:
            self._resize()

    def get(self, key: Any, default: Any = None) -> Any:
        """O(1) average; O(chain length) in the worst case."""
        for existing_key, value in self._buckets[self._index(key)]:
            if existing_key == key:
                return value
        return default

    def remove(self, key: Any) -> bool:
        """O(1) average. Returns whether the key was present."""
        chain = self._buckets[self._index(key)]
        for i, (existing_key, _) in enumerate(chain):
            if existing_key == key:
                chain.pop(i)
                self._size -= 1
                return True
        return False

    def __contains__(self, key: Any) -> bool:
        return any(k == key for k, _ in self._buckets[self._index(key)])

    def __len__(self) -> int:
        return self._size

    def __iter__(self) -> Iterator[tuple[Any, Any]]:
        for chain in self._buckets:
            yield from chain

    def keys(self) -> list[Any]:
        return [k for k, _ in self]

    def longest_chain(self) -> int:
        """Diagnostic: how badly is this table colliding?"""
        return max((len(chain) for chain in self._buckets), default=0)


class HashSet:
    """A set is a map whose values are ignored - so build it on one."""

    def __init__(self, items: list[Any] | None = None) -> None:
        self._map = HashMap()
        for item in items or []:
            self.add(item)

    def add(self, item: Any) -> None:
        self._map.put(item, True)

    def remove(self, item: Any) -> bool:
        return self._map.remove(item)

    def __contains__(self, item: Any) -> bool:
        return item in self._map

    def __len__(self) -> int:
        return len(self._map)


# ============================================================================
# 2. Frequency map
# ============================================================================
def char_frequency(s: str) -> dict[str, int]:
    """One pass, O(n). defaultdict avoids the "if key not in" boilerplate."""
    freq: dict[str, int] = defaultdict(int)
    for ch in s:
        freq[ch] += 1
    return dict(freq)


def first_unique_char(s: str) -> int:
    """Index of the first non-repeating character, or -1. Two passes, O(n)."""
    freq = Counter(s)
    for i, ch in enumerate(s):
        if freq[ch] == 1:
            return i
    return -1


# ============================================================================
# 3. Complement lookup
# ============================================================================
def two_sum(nums: list[int], target: int) -> tuple[int, int] | None:
    """Indices of the pair summing to target. O(n) time and space.

    The brute force asks "does x pair with any later element?" (O(n^2)).
    Hashing flips it: "has the complement already been seen?" - O(1) per check.
    """
    seen: dict[int, int] = {}                  # value -> index
    for i, x in enumerate(nums):
        if target - x in seen:
            return seen[target - x], i
        seen[x] = i                            # store AFTER checking: no reuse
    return None


def contains_duplicate(nums: list[int]) -> bool:
    """O(n) with a set, versus O(n log n) by sorting."""
    seen: set[int] = set()
    for x in nums:
        if x in seen:
            return True
        seen.add(x)
    return False


# ============================================================================
# 4. Grouping by a computed key
# ============================================================================
def group_anagrams(words: list[str]) -> list[list[str]]:
    """Group words that are anagrams. O(n * k log k) with a sorted-word key.

    Any function that maps equivalent items to the same value works as a key -
    that is the whole grouping pattern.
    """
    groups: dict[str, list[str]] = defaultdict(list)
    for word in words:
        groups["".join(sorted(word))].append(word)
    return list(groups.values())


# ============================================================================
# 5. Prefix sum + map
# ============================================================================
def subarray_sum_equals_k(nums: list[int], k: int) -> int:
    """Number of contiguous subarrays summing to k. O(n) time and space.

    prefix[j] - prefix[i] == k means the subarray (i, j] sums to k. So at each
    j, count how many earlier prefixes equal prefix[j] - k.

    counts[0] = 1 seeds the empty prefix, which is what makes subarrays that
    start at index 0 count.
    """
    counts: dict[int, int] = defaultdict(int)
    counts[0] = 1
    prefix = total = 0
    for x in nums:
        prefix += x
        total += counts[prefix - k]
        counts[prefix] += 1
    return total


# ============================================================================
# 6. Seen set
# ============================================================================
def longest_consecutive(nums: list[int]) -> int:
    """Longest run of consecutive integers. O(n), NOT O(n log n).

    Only start counting at a number whose predecessor is absent - so each run
    is walked exactly once and the whole thing stays linear despite the inner
    while loop.
    """
    unique = set(nums)
    best = 0
    for x in unique:
        if x - 1 in unique:
            continue                           # not the start of a run
        length = 1
        while x + length in unique:
            length += 1
        best = max(best, length)
    return best


def intersection(a: list[int], b: list[int]) -> list[int]:
    """Common elements. O(n + m) with a set; O(n * m) with `in list`."""
    lookup = set(b)
    return sorted({x for x in a if x in lookup})


# ============================================================================
# 7. Hashing + ordering: an LRU cache
# ============================================================================
class LRUCache:
    """O(1) get and put with a capacity limit.

    Needs both a hash map (O(1) lookup) and an ordering structure (to know
    what is least recently used). OrderedDict is a dict plus a doubly linked
    list, which is exactly that combination.
    """

    def __init__(self, capacity: int) -> None:
        if capacity <= 0:
            raise ValueError("capacity must be positive")
        self._capacity = capacity
        self._data: OrderedDict[Any, Any] = OrderedDict()

    def get(self, key: Any) -> Any:
        if key not in self._data:
            return -1
        self._data.move_to_end(key)            # mark as most recently used
        return self._data[key]

    def put(self, key: Any, value: Any) -> None:
        if key in self._data:
            self._data.move_to_end(key)
        self._data[key] = value
        if len(self._data) > self._capacity:
            self._data.popitem(last=False)     # evict the least recent

    def keys_in_order(self) -> list[Any]:
        """Least recently used first."""
        return list(self._data.keys())


# ============================================================================
# demo
# ============================================================================
def demo() -> None:
    m = HashMap()
    m.put("cat", 1)
    m.put("dog", 2)
    m.put("cat", 9)                            # overwrite, not duplicate
    assert len(m) == 2 and m.get("cat") == 9
    assert m.get("missing") is None and m.get("missing", 0) == 0
    assert "dog" in m and "bird" not in m
    assert m.remove("dog") and not m.remove("dog")
    assert len(m) == 1

    # Resizing: insert enough keys to force at least one growth, and check
    # that every key survives the rehash.
    big = HashMap()
    for i in range(100):
        big.put(f"key{i}", i)
    assert len(big) == 100
    assert big.capacity > HashMap._INITIAL_CAPACITY      # it grew
    assert big.load_factor <= HashMap._MAX_LOAD_FACTOR   # and stayed healthy
    assert all(big.get(f"key{i}") == i for i in range(100))
    assert big.longest_chain() <= 5                      # collisions stay tame
    assert sorted(big.keys()) == sorted(f"key{i}" for i in range(100))

    # Tuples are hashable and work as compound keys; lists do not.
    coords = HashMap()
    coords.put((1, 2), "origin-ish")
    assert coords.get((1, 2)) == "origin-ish"

    s = HashSet([1, 2, 2, 3])
    assert len(s) == 3 and 2 in s and 9 not in s
    assert s.remove(2) and 2 not in s

    assert char_frequency("aab") == {"a": 2, "b": 1}
    assert first_unique_char("leetcode") == 0
    assert first_unique_char("aabb") == -1

    assert two_sum([2, 7, 11, 15], 9) == (0, 1)
    assert two_sum([3, 3], 6) == (0, 1)         # duplicates, distinct indices
    assert two_sum([1, 2], 99) is None
    assert contains_duplicate([1, 2, 3, 1])
    assert not contains_duplicate([1, 2, 3])

    groups = group_anagrams(["eat", "tea", "tan", "ate", "nat", "bat"])
    assert sorted(len(g) for g in groups) == [1, 2, 3]

    assert subarray_sum_equals_k([1, 1, 1], 2) == 2
    assert subarray_sum_equals_k([1, 2, 3], 3) == 2      # [1,2] and [3]
    assert subarray_sum_equals_k([1, -1, 0], 0) == 3     # negatives work too

    assert longest_consecutive([100, 4, 200, 1, 3, 2]) == 4    # 1,2,3,4
    assert longest_consecutive([]) == 0
    assert intersection([1, 2, 2, 1], [2, 2]) == [2]

    lru = LRUCache(2)
    lru.put("a", 1)
    lru.put("b", 2)
    assert lru.get("a") == 1                    # "a" is now the most recent
    lru.put("c", 3)                             # evicts "b", the least recent
    assert lru.get("b") == -1
    assert lru.keys_in_order() == ["a", "c"]

    print("10-Hashing (Python): all checks passed")


if __name__ == "__main__":
    demo()
